### PR TITLE
Add Revert mapping for matrix row/column ids

### PR DESCRIPTION
**Pull request by NicholasCouri**
The idea is to have an initial version of the reverse mapping for identifying rows and columns from their ids.

We update the rowMap/colMap whenever Matrix ops are generated - The tricky part is for the scenario where the channel is just about to be created for the first time (we get a channelOp) - in this case we would have the matrix but not the rowMap/rowcolumn. So, on that first time, we  populate the entire rowMap/rowColumn and pay this cost. 
From that moment on, the matrix operations ensure the rowMap and colMap are going to be updated correctly. 

I plan on looking into on the Tablero implementation and, also, on adding new tests to exercise row/column manipulation.

### CODE REVIEW

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 1]
```diff
@@ -455,42 +457,41 @@ export class TempCollabSpaceRuntime
...
-			if ((this.matrix.getCell(0, col) as unknown as string) === colId) {
-				break;
-			}
-		}
-		assert(col !== colCount, "channel not found");
+		const row = this.rowMap.get(rowId);
+		const col = this.colMap.get(colId);
 
+		assert(row !== undefined, "channel's row not found");
+		assert(col !== undefined, "channel's col not found");
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: Please add validation that indeed cells[row, 0] === rowId, same for column.
It's a bit expensive (2 more reads from matrix, no easy way to produce debug only code in JS), but I'd rather pay that price now (and remove it later if profiles point to it) then to be sorry we did not find a bug sooner. #Resolved

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 2]
```diff
@@ -666,12 +667,18 @@ export class TempCollabSpaceRuntime
 		// generate new ID for a columns
 		while (count > 0) {
 			count--;
-			this.matrix.setCell(0, col, uuid() as unknown as MatrixInternalType);
+			const colId = uuid();
+			this.matrix.setCell(0, col, colId as unknown as MatrixInternalType);
+			this.colMap.set(colId, col);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: The problem with doing it here is that it only works for local changes.
We need to ensure we update mappings on both local & remote changes.
For that reason, it would be better if such updates where in response to events that fire in response to changes, and thus capture both local & remote changes

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 3]
```diff
@@ -455,42 +457,41 @@ export class TempCollabSpaceRuntime
...
-		const colCount = this.matrix.colCount;
-
-		let row;
-		for (row = 1; row < rowCount; row++) {
-			if ((this.matrix.getCell(row, 0) as unknown as string) === rowId) {
-				break;
-			}
+		// In case the channel is about to be created we need to populate the rowMap and colMap maps.
+		if (this.rowMap.size === 0 || this.colMap.size === 0) {
+			this.populateRowColMaps();
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: That's probably fine to do so on demand but looks a bit strange. Minor thing - a bit expensive if table is empty (very unlikely to get here with empty table, but possible).

Why not do it in loadCore()?

**Comment by vladsud** [comment]
BTW, I'd recommend you to start with adding UT, including concurrent UTs, and ensure that it catches bugs (i.e. there are enough validation in shipping and test code to catch the fact that IDs are out of order).
While shipping (production) code should do most of the work (such that we can find really hard to find bugs in production), feel free to also add test-only methods on IEfficientMatrixTest and use them in test to poke into internal structure of data.

File: packages/framework/collabspaces/src/contracts.ts [hunk 4]
```diff
@@ -84,4 +84,16 @@ export interface IEfficientMatrixTest {
 
 	// Returns a structure with various debug info about the cell
 	getCellDebugInfo(row: number, col: number): Promise<{ channel?: ICollabChannelCore }>;
+
+	getCellChannelIdDebugInfo(row: number, col: number): Promise<{ channelId: string }>;
```
Comment (vladsud) [inline packages/framework/collabspaces/src/contracts.ts]: I'd roll it into single API getCellDebugInfo() - All the info about cell.
 #Resolved

File: packages/framework/collabspaces/src/contracts.ts [hunk 5]
```diff
@@ -84,4 +84,16 @@ export interface IEfficientMatrixTest {
...
+
+	getReverseMapsDebugInfo(
+		rowId: string,
+		colId: string,
+	): {
```
Comment (vladsud) [inline packages/framework/collabspaces/src/contracts.ts]: I'm not sure returning rowMapSize & colMapSize is useful - I'd rather add asserts across all of the code (including this API call) that they are the same as size of the matrix,
I'd also probably not use word "Index" as we do not use it on other APIs - a bit easier to connect the dots.

File: packages/framework/collabspaces/src/test/collabSpaces.spec.ts [hunk 6]
```diff
@@ -494,12 +593,14 @@ describe("Temporal Collab Spaces", () => {
 		// such conversion works correctly.
 	}
 
-	it("Channel overwrite with syncronization", async () => {
-		await ChannelOverwrite(true);
-	});
+	describe("Channel Overwrite", () => {
```
Comment (vladsud) [inline packages/framework/collabspaces/src/test/collabSpaces.spec.ts]: You are a bit behind on changes - please merge latest changes (there are more tests and more arguments here #Resolved

File: packages/framework/collabspaces/src/test/collabSpaces.spec.ts [hunk 7]
```diff
@@ -309,6 +310,104 @@ describe("Temporal Collab Spaces", () => {
 		assert(value2?.value === initialValue, "value was preserved correctly");
 	});
 
+	describe("Reverse Mapping tests", () => {
+		async function getRowAndColIdsFromChannelId(
+			collabSpace: IEfficientMatrix & IEfficientMatrixTest,
+			row: number,
+			col: number,
+		): Promise<{ rowId: string; colId: string }> {
+			const result = await collabSpace.getCellChannelIdDebugInfo(row, col);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/test/collabSpaces.spec.ts]: I'd suggest to do this kind of parsing in one place (inside of mapChannelToCell()) and simply for getCellChannelIdDebugInfo or other API to return all the required info. That way it's a bit easier to change it if we need to in the future #Resolved

File: packages/framework/collabspaces/src/test/collabSpaces.spec.ts [hunk 8]
```diff
@@ -309,6 +310,104 @@ describe("Temporal Collab Spaces", () => {
...
+			const col = 3;
+			const { rowId, colId } = await getRowAndColIdsFromChannelId(collabSpace, row, col);
+
+			const { rowMapSize, colMapSize, rowIndex, colIndex } =
+				collabSpace.getReverseMapsDebugInfo(rowId, colId);
+
+			assert(rowMapSize === rows, "rowMapSize is incorrect");
+			assert(rowIndex === row + 1, "rowIndex from the actual matrix has to be offset by 1");
+			assert(colMapSize === cols, "colMapSize is incorrect");
+			assert(colIndex === col + 1, "colIndex is correct");
```
Comment (vladsud) [inline packages/framework/collabspaces/src/test/collabSpaces.spec.ts]: I'd suggest for API surface to be consistent. I.e. for getReverseMapsDebugInfo() to do -1, and thus all the code (including test code) not to deal with +1 or -1. We could change (in the future) where and how we store this data, it would be great (even for test code) not to depend on a fact that we have an extra row / col

File: packages/framework/collabspaces/src/contracts.ts (hunk 5 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/contracts.ts]: My goal was to automatically tell whether the reverse mapping matrixes were updated using their size as a reference.

File: packages/framework/collabspaces/src/test/collabSpaces.spec.ts [hunk 9]
```diff
@@ -309,6 +310,104 @@ describe("Temporal Collab Spaces", () => {
...
+			const { rowId, colId } = await getRowAndColIdsFromChannelId(collabSpace, row, col);
+
+			assert(nextColId === colId, "colId after removal should be the same as nextColId");
+
+			const { rowMapSize, colMapSize, rowIndex, colIndex } =
+				collabSpace.getReverseMapsDebugInfo(rowId, colId);
+			assert(rowMapSize === rows, "rowMapSize is incorrect");
+			assert(rowIndex === row + 1, "rowIndex is correct");
+			assert(colMapSize === cols - columnsToBeRemoved, "colMapSize is incorrect");
+			assert(colIndex === col + 1, "colIndex is correct");
```
Comment (vladsud) [inline packages/framework/collabspaces/src/test/collabSpaces.spec.ts]: I'd suggest to have at least one test that deals with changes in another container, and validates that mappings are Ok in both containers after changes are synced. Including having concurrent changes (like insert a row in both), similar to what we have with ChannelOverwrite() (with parametrizing of having synchronization vs. not, and thus testing concurrency vs. non-concurrent change propagation)

Comment (NicholasCouri) [reply packages/framework/collabspaces/src/test/collabSpaces.spec.ts]: yes, I plan on doing it but I wanted to see if there is a better way to store and update the reverse mapping matrixes  for sure.

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 10]
```diff
@@ -415,9 +417,35 @@ export class TempCollabSpaceRuntime
...
+					map.set(key, index + 1);
+				});
+			}
+		};
+
 		this.matrix.openMatrix({
-			rowsChanged: (rowStart: number, removedCount: number, insertedCount: number) => {},
-			colsChanged: (colStart: number, removedCount: number, insertedCount: number) => {},
+			rowsChanged: (rowStart: number, removedCount: number, insertedCount: number) => {
+				removeCellsFromMap(rowStart, removedCount, this.rowMap);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: We get here for both cases - rows added, as well as rows deleted. I think the call as it stands only deals with removals. #Resolved

Comment (NicholasCouri) [reply packages/framework/collabspaces/src/collabSpaces.ts]: rows added do not help as we cant add blank rows on the row/column Maps

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 11]
```diff
@@ -415,9 +417,35 @@ export class TempCollabSpaceRuntime
...
+		// remove the tracking information as well as update the indexes in the map.
+		const removeCellsFromMap = (start: number, count: number, map: Map<string, any>) => {
+			if (count > 0) {
+				// Convert the map to an array of key-value pairs
+				const entries = Array.from(map.entries());
+
+				// Remove the specified range of items from the array
+				// Note the index from the reverse mapping is off by 1 as the matrix has a row and col tracking IDs
+				// and the map indexes are 0 based
+				entries.splice(start - 2, count);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: I'm not sure how to think about order here. when you convert a map into array, what dictates the order of the entries in array? The values will be inserted randomly into this map (as user adds and deletes rows). So it sounds like the order here is simply random, and I do not think we can assume anything.

Before fixing anything, can you please add a unit test that hits that problem? Such that we have coverage for future? #Resolved

File: packages/framework/collabspaces/src/collabSpaces.ts (hunk 10 above)
Comment (vladsud) [reply packages/framework/collabspaces/src/collabSpaces.ts:444]: Or is logic in insertCellIntoReverseMap() tries to enforce some order? If it does, are you sure that map.entries() provides us some guarantees? I do not see anything listed on https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries, though I have not looked at full spec

File: packages/framework/collabspaces/src/collabSpaces.ts (hunk 11 above)
Comment (vladsud) [reply packages/framework/collabspaces/src/collabSpaces.ts]: Or is logic in insertCellIntoReverseMap() tries to enforce some order? If it does, are you sure that map.entries() provides us some guarantees? I do not see anything listed on https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map/entries, though I have not looked at full spec

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 12]
```diff
@@ -439,25 +467,77 @@ export class TempCollabSpaceRuntime
...
-			const channelId = info.channelId;
-			// Need to update channel and convert it to real thing.
-			const deferredChannel = this.deferredChannels.get(channelId);
-			assert(deferredChannel !== undefined, "deferred channel not found");
-			this.deferredChannels.delete(channelId);
 
-			this.destroyChannelCore(channelId);
+		if (info.value !== undefined) {
+			// Col and Row == -1 means cell that uniquely identify the rows and columns from the matrix.
+			this.addCellToReverseMap(row, col, info);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: I see some code that I added earlier as green (as addition) - looks like the merge was not properly done? Can you please merge latest changes? #Resolved

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 13]
```diff
@@ -439,25 +467,77 @@ export class TempCollabSpaceRuntime
...
+					row,
+					info.value as unknown as string,
+					row + 1,
+				);
+			} else {
+				this.rowMap.set(info.value as unknown as string, row + 1);
+			}
+		}
+		if (row === -1) {
+			this.colMap.set(info.value as unknown as string, col + 1);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: I'd expect code to be symmetrical for rows & cols. I'm not sure why col === - 1 case has more sub-cases compared to row === -1 path. #Resolved

File: packages/framework/collabspaces/src/collabSpaces.ts (hunk 10 above)
Comment (vladsud) [reply packages/framework/collabspaces/src/collabSpaces.ts]: Oh, I see, the issue with inserts - we expect to get IDs from remote clients. Maybe at least add some comments on that, as it's a bit hard to follow.

File: packages/framework/collabspaces/src/collabSpaces.ts (hunk 11 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/collabSpaces.ts]: It seems I would need to use Object.entries() to generate the array - as it keeps the order from the Map. I will add more testing  and also look at the Tablero implementation as I mentioned.

File: packages/framework/collabspaces/src/collabSpaces.ts (hunk 13 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/collabSpaces.ts]: the problem once again is the insertion of rows - we get one by one - I'll add a comment to make it clear.
removal of columns within existing boundaries are dealt by rowsChanged and columnsChanged methods

File: packages/framework/collabspaces/src/contracts.ts [hunk 14]
```diff
@@ -83,5 +94,27 @@ export interface IEfficientMatrixTest {
...
-	getCellDebugInfo(row: number, col: number): Promise<{ channel?: ICollabChannelCore }>;
+	getCellDebugInfo(
+		row: number,
+		col: number,
+	): Promise<{ channel?: ICollabChannelCore; channelId: string; rowId: string; colId: string }>;
+
+	getReverseMapsDebugInfo(
+		rowId?: string | undefined,
+		colId?: string | undefined,
+	): ReverseMapsDebugInfoType;
```
Comment (vladsud) [inline packages/framework/collabspaces/src/contracts.ts]: API looks a bit weird - it returns data about specific cell, but also about whole matrix. I'd break it into two APIs.

File: packages/framework/collabspaces/src/contracts.ts [hunk 15]
```diff
@@ -83,5 +94,27 @@ export interface IEfficientMatrixTest {
...
+}
+
+export type ReverseMapType = "row" | "col";
+
+export interface IReverseMap {
```
Comment (vladsud) [inline packages/framework/collabspaces/src/contracts.ts]: I believe this should not be here - this file is about public contracts.
Maybe in the future we would benefit from having "internal" contracts file (or multiple implementations of this interface), but for now I'd just move it into reverseMap.ts

File: packages/framework/collabspaces/src/index.ts [hunk 16]
```diff
@@ -11,5 +11,6 @@ export {
 	ICollabChannelCore,
 	ICollabChannel,
 	CollabSpaceCellType,
+	ReverseMapsDebugInfoType,
```
Comment (vladsud) [inline packages/framework/collabspaces/src/index.ts]: Why? This is becoming public API. Tests do not require that - they can directly get to the guts of the implementation.

File: packages/framework/collabspaces/src/contracts.ts (hunk 15 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/contracts.ts]: yes, moved to reverseMap.ts for now

File: packages/framework/collabspaces/src/contracts.ts (hunk 14 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/contracts.ts]: split into 2

File: packages/framework/collabspaces/src/index.ts (hunk 16 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/index.ts]: Because it was used within  contract.ts - inside the IEfficientMatrixTest  which is also exported in  index.ts. 
Anyways, now that we split the implementation of getReverseMapsDebugInfo, I removed the type for now.

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 17]
```diff
@@ -447,23 +457,42 @@ export class TempCollabSpaceRuntime
...
-		if (info.value !== undefined && isChannelDeffered(info.channelInfo?.type)) {
-			const channelId = info.channelId;
-			// Need to update channel and convert it to real thing.
-			const deferredChannel = this.deferredChannels.get(channelId);
-			assert(deferredChannel !== undefined, "deferred channel not found");
-			this.deferredChannels.delete(channelId);
 
-			this.destroyChannelCore(channelId);
+		if (info.value !== undefined) {
+			if (col === -1 || row === -1) {
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: This is more of my mistake but would be great if you could fix it.
Calling getCellInfo() above with either row or col being -1 is not great. It's great that it does not crash today, but it could, and it also does a bunch of unnecessary work in such case.
I'd advice to actually not even get into here with -1, but make a fork much earlier - in colsChanged callback and have separate path for -1 that would not call this.getCellInfo, but would call minimum APIs (I assume go directly against this.matrix.getCell) #Resolved

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 18]
```diff
@@ -502,9 +531,35 @@ export class TempCollabSpaceRuntime
 	public async getCellDebugInfo(
 		row: number,
 		col: number,
-	): Promise<{ channel?: ICollabChannelCore }> {
+	): Promise<{ channel?: ICollabChannelCore; channelId?: string; rowId: string; colId: string }> {
 		const result = this.getCellInfo(row, col);
-		return { channel: await result.channel };
+		assert(result.channelId !== undefined, "channelId is missing");
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: This would assert if called for a cell that has no channel. From API name / semantics it does not make much sense. #Resolved

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 19]
```diff
@@ -502,9 +531,35 @@ export class TempCollabSpaceRuntime
...
+		const channel = await result.channel;
+		return { channel, channelId: result.channelId, rowId, colId };
+	}
+
+	public getReverseMapCellDebugInfo(rowId: string, colId: string): { row: number; col: number } {
+		const row = this.reverseMap.getRowId(rowId);
+		const col = this.reverseMap.getColId(colId);
+		assert(row !== undefined, "rowIndex is missing");
+		assert(col !== undefined, "colIndex is missing");
+		return {
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts:646]: Given this is debug API, I'd expect it to immediately do reverse call (through getCellDebugInfo()?) and ensure that we can map row/col back to exactly same rowId, colId. #Resolved

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 20]
```diff
@@ -502,9 +531,35 @@ export class TempCollabSpaceRuntime
...
 		const result = this.getCellInfo(row, col);
-		return { channel: await result.channel };
+		assert(result.channelId !== undefined, "channelId is missing");
+		const { rowId, colId } = this.parseChannelId(result.channelId);
+		const channel = await result.channel;
+		return { channel, channelId: result.channelId, rowId, colId };
+	}
+
+	public getReverseMapCellDebugInfo(rowId: string, colId: string): { row: number; col: number } {
+		const row = this.reverseMap.getRowId(rowId);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: API name does not make sense. It returns index, not rowId, so it should not be called getRowId. Same for getColId #Resolved

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 21]
```diff
@@ -502,9 +531,35 @@ export class TempCollabSpaceRuntime
...
+		return {
+			row: row - 1,
+			col: col - 1,
+		};
+	}
+
+	// For test purposes only!
+	// Returns the Reverse Map size and the actual indexes from matrix stored on the reverse mapping matrixes.
+	public getReverseMapsDebugInfo(): {
+		rowMap: { [id: string]: number };
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts:655]: Consider wrapping return type with Readonly<>, such that caller can't change it. #Resolved

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 22]
```diff
@@ -573,38 +628,30 @@ export class TempCollabSpaceRuntime
...
+	private mapChannelToCell(channelId: string): { row: number; col: number; iteration: string } {
+		const { rowId, colId, iteration } = this.parseChannelId(channelId);
+		const row = this.reverseMap.getRowId(rowId);
+		const col = this.reverseMap.getColId(colId);
 
+		assert(
+			row !== undefined &&
+				this.areEqualUuid(this.matrix.getCell(row, 0) as unknown as uuidType, rowId),
+			"channel's row not found",
+		);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts:746]: I was wrong here - please see latest proposed changes in https://github.com/vladsud/FluidFramework/pull/13/files
I do not care in what order we merge our PRs (I can fix it), but this function needs to deal properly (not crash or assert) when it can't map channel to cell.

File: packages/framework/collabspaces/src/reverseMap.ts [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+	removeCellsFromMap(type: ReverseMapType, start: number, count: number): void;
+
+	addCellToMap(type: ReverseMapType, id: string, index: number): void;
+}
+
+export class ReverseMap implements IReverseMap {
+	private readonly rowMap: { [id: string]: number } = {};
+	private readonly colMap: { [id: string]: number } = {};
+
+	constructor() {}
```
Comment (vladsud) [inline packages/framework/collabspaces/src/reverseMap.ts]: Maybe I'm missing something, but this should fail as you do not populate initial state after loading matrix.
It would work only when we create it from scratch, not when we load existing matrix.

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 24]
```diff
@@ -447,23 +457,42 @@ export class TempCollabSpaceRuntime
...
-			assert(deferredChannel !== undefined, "deferred channel not found");
-			this.deferredChannels.delete(channelId);
 
-			this.destroyChannelCore(channelId);
+		if (info.value !== undefined) {
+			if (col === -1 || row === -1) {
+				// Col and Row == -1 means cell that uniquely identify the rows and columns from the matrix.
+				const mapInfo: { type: ReverseMapType; index: number } =
+					col === -1 ? { type: "row", index: row } : { type: "col", index: col };
+				this.reverseMap.addCellToMap(
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: Let's add Pri1 comment in the code here to revisit how we raise matrix events.
Today we will raise "column added" event before we have an ID, and thus exposing incorrect state of matrix to consumers.
It's better if we do something like that (as a separate PR):
- suppress col/row addition events
- raise them here, when we learn about new ID.
It's not that simple, as proposed sequence would only work for a single row/col addition, not for many rows/columns added, but I'm sure we can figure out something.
Alternatively, we need better APIs from matrix... #Resolved

File: packages/framework/collabspaces/src/reverseMap.ts [hunk 25]
```diff
@@ -0,0 +1,108 @@
...
+
+	public getRowId(rowId: string): number | undefined {
+		return this.rowMap[rowId];
+	}
+
+	public getColId(colId: string): number | undefined {
+		return this.colMap[colId];
+	}
+
+	public getRowMap(): { [id: string]: number } {
```
Comment (vladsud) [inline packages/framework/collabspaces/src/reverseMap.ts]: Please make return types Readonly<> #Resolved

File: packages/framework/collabspaces/src/reverseMap.ts [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+		const map = type === "row" ? this.rowMap : this.colMap;
+		if (count > 0) {
+			// Convert the map to an array of key-value pairs
+			// Notice we use Object.entries to keep the order from the map
+			const entries = Object.entries(map);
+
+			// Remove the specified range of items from the array
+			// Note the index from the reverse mapping is off by 1 as the matrix has a row and col tracking IDs
+			// and the map indexes are 0 based
+			entries.splice(start - 1, count);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/reverseMap.ts]: I'd suggest this code to have 0-based indexing, and deal with -1 thing in the caller.

File: packages/framework/collabspaces/src/reverseMap.ts [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+
+			// Clear the original map
+			Object.keys(map).forEach((key) => {
+				if (Object.prototype.hasOwnProperty.call(map, key)) {
+					Reflect.deleteProperty(map, key);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/reverseMap.ts]: We should measure it, but from what I know, deleting a property is very expensive operation on object. Probably does not matter for our scale, but worth checking performance of it if we are dealing with 10K rows being deleted. #Pending

File: packages/framework/collabspaces/src/reverseMap.ts [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+
+		// Re-populate the map with the updated items from the array
+		entries.forEach(([localKey], newIndex) => {
+			map[localKey] = newIndex + 1;
+		});
+	}
+
+	public addCellToMap(type: ReverseMapType, id: string, index: number): void {
+		const map = type === "row" ? this.rowMap : this.colMap;
+		// if (type === "row") {
```
Comment (vladsud) [inline packages/framework/collabspaces/src/reverseMap.ts]: delete this line? #Resolved

File: packages/framework/collabspaces/src/reverseMap.ts [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+		// without the unique ids and soon after, the cell changes are triggered and we take opportunity to update the reverse map.
+		if (Object.keys(map).length >= index + 1) {
+			// In case of insertion or rows within existing boundaries of the matrix, we will need to manually update
+			// the map to reflect and shift existing indexes. For example, if we insert a new row at index 2,
+			// the existing row at index 2 will be shifted to index 3 and so on.
+			// Note this only applies for insertion within existing rows but removal as
+			//  rowsChanged and columnsChanged methods are responsible for processing it.
+			this.insertCellIntoMap(type, index, id, index + 1);
+		} else {
+			map[id] = index + 1;
```
Comment (vladsud) [inline packages/framework/collabspaces/src/reverseMap.ts]: similar feedback as below - I'd use 0-based indexes (i.e. not do +1 here), and let consumer code do +1 on it after retrieving back index from this object. That way we are not splicing responsibilities - only caller deals with +/- 1 logic, and this code deals with zero-based indexes.

File: packages/framework/collabspaces/src/test/collabSpaces.spec.ts [hunk 30]
```diff
@@ -319,6 +329,201 @@ describe("Temporal Collab Spaces", () => {
...
+					const colValue = map2[id];
+					assert(
+						rowValue === colValue,
+						`Values for id ${id} do not match: row = ${rowValue}, col = ${colValue}`,
+					);
+				} else {
+					assert(false, `No matching id found in map2 for ${id}`);
+				}
+			}
+			return true;
```
Comment (vladsud) [inline packages/framework/collabspaces/src/test/collabSpaces.spec.ts]: function should return void - you never check its return value, and it's always true, so it's useless #Resolved

File: packages/framework/collabspaces/src/test/collabSpaces.spec.ts [hunk 31]
```diff
@@ -383,6 +588,11 @@ describe("Temporal Collab Spaces", () => {
 		// recreate deleted channel
 		channel = (await collabSpace.getCellChannel(row, col)) as ISharedCounter;
 
+		// TODO: (nichoc) investigate as without the following lines, the test fails.
```
Comment (vladsud) [inline packages/framework/collabspaces/src/test/collabSpaces.spec.ts]: Let's not leave it as is - we need to better understand what happens here.
If you are Ok to pay a merge cost, I can merge my latest PR that add a bunch of stress tests and fixes some problems, and once we merge latest into this PR, I can also take a look at it. #Resolved

File: packages/framework/collabspaces/src/reverseMap.ts (hunk 23 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/reverseMap.ts]: whenever any cell gets changed/added the callbacks populate it. So we do not need to do anything here.

File: packages/framework/collabspaces/src/test/collabSpaces.spec.ts [hunk 32]
```diff
@@ -705,34 +940,56 @@ describe("Temporal Collab Spaces", () => {
...
+			cp.insertCols(pos, count);
+			currentColCount += count;
+			commandArray.push(
+				`insertColsFn post pos ${pos}, count ${count}, cp.colCount ${cp.colCount}  RowCount ${currentRowCount}, ColCount ${currentColCount}`,
+			);
 		};
 
 		const insertRowsFn: Op = async (cp: IMatrix) => {
-			cp.insertRows(rand(cp.rowCount), 1 + rand(3));
+			const pos = rand(Math.min(currentRowCount, cp.colCount));
```
Comment (vladsud) [inline packages/framework/collabspaces/src/test/collabSpaces.spec.ts]: I do not think this is correct. In a sense that - this limits our test coverage.
It's totally normal for 2 containers to have different view on how many rows are there in container while they did not synchronize their changes. If this improves perf, then likely we have a bug in matrix that needs to be looked at. Based on what I understand, there should be no perf difference as Sequence (used as implementation for permutation vector) should not care and should have roughly same perf enumerating deleted segments as it does for "real" segments.
You might want to follow up with DanielL on that.

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 33]
```diff
@@ -496,8 +523,16 @@ export class CollabSpacesRuntime
...
-						// -1 due to first row & col tracking IDs
-						if (row !== 0 && col !== 0) {
+						if (row === 0 || col === 0) {
+							// TBD(Pri1) : Today we will raise "column added" event before we have an ID, and thus exposing incorrect state of matrix to consumers.
+							// It's better if we do something like that:
+							// - suppress col/row addition events
+							// - raise them here, when we learn about new ID.
+							// It's not that simple, as proposed sequence would only work for a single row/col addition, not for many rows/columns added, but I'm sure we can figure out something.
+							// Alternatively, we need better APIs from matrix...
+							this.updateReverseMap(row - 1, col - 1);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts:533]: This is probably fine, but API looks a bit weird. I.e. we pass -1 for one of the parameters and that tells it what to update.
Maybe better to have that logic here and call two different APIs (updateColReverseMap, updateRowReverseMap) or single API (updateMap(what: "rows" | "cols", index).

Comment (vladsud) [reply packages/framework/collabspaces/src/collabSpaces.ts:533]: Ah, you are doing it in updateReverseMap - that's probably fine, but please see another comment RE -1 logic below

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 34]
```diff
@@ -511,8 +546,23 @@ export class CollabSpacesRuntime
 		return this.matrixInternal;
 	}
 
+	private updateReverseMap(row: number, col: number) {
+		// the first row and column are used to track the internal IDs
+		const info = this.getCellInfo(row, col);
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: Shouldn't this fail? I.e. we are calling something like getCellInfo(5, -1).
I'd suggest calling directly matrix.getCell (and thus not do -1 in the caller of this API), and only once you got everything, call this.reverseMap.addCellToMap() and do -1 logic right before this call

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 35]
```diff
@@ -537,7 +587,7 @@ export class CollabSpacesRuntime
 	private getCellInfo(rowArg: number, colArg: number): ICellInfo {
 		const row = rowArg + 1;
 		const col = colArg + 1;
-		assert(row > 0 && col > 0, "arguments");
+		assert(row >= 0 && col >= 0, "arguments");
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: I'd not make this change - see comments above. All you need in updateReverseMap() is getting the value from matrix - call it matrix directly. This function is doing a lot of work that is not needed for your case, and it's better for it to be specialized to work with "normal" cells, not mapping cells

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 36]
```diff
@@ -569,9 +619,46 @@ export class CollabSpacesRuntime
...
+		channelId?: string;
+		rowId?: string;
+		colId?: string;
+	}> {
 		const result = this.getCellInfo(row, col);
-		return { channel: await result.channel };
+		const { rowId = undefined, colId = undefined } =
+			result.channelId !== undefined ? this.parseChannelId(result.channelId) : {};
+		const channel = await result.channel;
+		return { channel, channelId: result.channelId, rowId, colId };
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: I wonder if you can do the following:

```ts
const info = result.channelId !== undefined ? this.parseChannelId(result.channelId);
const channel = await result.channel;
return { channel, ...info};
```
it returns a bit more info (iteration), but that should not matter. Feels a bit more readable. 
(this assumes that tsc is happy with it)

File: packages/framework/collabspaces/src/collabSpaces.ts (hunk 35 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/collabSpaces.ts]: good point.

File: packages/framework/collabspaces/src/collabSpaces.ts [hunk 37]
```diff
@@ -636,42 +723,35 @@ export class CollabSpacesRuntime
...
-		// or move most of the channel logic to the layer that implements row/col stable IDs.
-
-		const rowCount = this.matrix.rowCount;
-		const colCount = this.matrix.colCount;
+	private mapChannelToCellCore(channelId: string) {
+		const { rowId, colId, iteration } = this.parseChannelId(channelId);
+		const row = this.reverseMap.getRowIndex(rowId);
+		const col = this.reverseMap.getColIndex(colId);
 
-		let row;
```
Comment (vladsud) [inline packages/framework/collabspaces/src/collabSpaces.ts]: Any chance we can have two implementations for remapping - the old logic and new logic? Such that we can easily switch back and forth when investigating issue?

Comment (vladsud) [reply packages/framework/collabspaces/src/collabSpaces.ts]: If it's too hard, then probably it's Ok not to do it,

Comment (NicholasCouri) [reply packages/framework/collabspaces/src/collabSpaces.ts]: sure thing. I was planning on leaving the remapping alive and simply do not make use of it when doing the mapChannelToCellCore in case this flag is off. How does it sound ? I could also disable it entirely if the flag is off but I suppose it might be helpful to have it around and debug things in the future.

Comment (NicholasCouri) [reply packages/framework/collabspaces/src/collabSpaces.ts]: Or do you want to turn it on or off before the test execution ? Like, run this test with the reverseMapping off and then on ? 
Like configure at the factory level ?  I was thinking on a simple global at the collabspace level.

File: packages/framework/collabspaces/src/test/collabSpaces.spec.ts (hunk 32 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/test/collabSpaces.spec.ts]: I will revert it for now and track it separately as it does not affect passing/failures.

File: packages/framework/collabspaces/src/collabSpaces.ts (hunk 36 above)
Comment (NicholasCouri) [reply packages/framework/collabspaces/src/collabSpaces.ts]: I do not have  a strong opinion either way  - I will change it.

**Review by vladsud — APPROVED** [review]
